firmware/application/ui/ui_freqlist.cpp

```diff
@@ -0,0 +1,169 @@
+ * along with this program; see the file COPYING.  If not, write to
+ * the Free Software Foundation, Inc., 51 Franklin Street,
+ * Boston, MA 02110-1301, USA.
+ */
+
+#include "ui_freqlist.hpp"
+
+#include "baseband_api.hpp"
+#include "utility.hpp"
+
+#include <algorithm>
+
+namespace ui {
+FreqManUIList::FreqManUIList(
+    Rect parent_rect,
+    bool instant_exec)
+    : Widget{parent_rect},
+      instant_exec_{instant_exec} {
+    this->set_focusable(true);
+}
+
+uint8_t FreqManUIList::set_highlighted_index(int index) {
+    if ((unsigned)(current_index + index) >= freqlist_db.size())
+        return highlighted_index;
+    if (freqlist_db.size() == 0) {
+        current_index = highlighted_index = 0;
+        return 0;
+    }
+    if (index < 0) {
+        index = 0;
+        if (current_index > 0)
+            current_index--;
+    }
+    if (index >= freqlist_nb_lines) {
+        index = freqlist_nb_lines - 1;
+        if ((unsigned)(current_index + index) < freqlist_db.size())
+            current_index++;
+        else
+            current_index = freqlist_db.size() - freqlist_nb_lines - 1;
+    }
+
+    highlighted_index = index;
+
+    return index;
+}
+
+uint8_t FreqManUIList::get_index() {
+    return current_index + highlighted_index;
+}
+
+void FreqManUIList::paint(Painter& painter) {
+    freqlist_nb_lines = 0;
+    const auto r = screen_rect();
+    uint8_t focused = has_focus();
+    const Rect r_widget_screen{r.left() + focused, r.top() + focused, r.width() - 2 * focused, r.height() - 2 * +focused};
+    painter.fill_rectangle(
+        r_widget_screen,
+        Color::black());
+    if (freqlist_db.size() == 0)
+        return;
+    uint8_t nb_lines = 0;
+    for (uint8_t it = current_index; it < freqlist_db.size(); it++) {
+        uint8_t line_height = (int)nb_lines * char_height;
+        if (line_height < (r.height() - char_height)) {
+            // line is within the widget
+            painter.draw_string(
+                {0, r.location().y() + (int)nb_lines * char_height},
+                style_default, freqman_item_string(freqlist_db[it], 30));
+            if (nb_lines == highlighted_index) {
+                const Rect r_highlighted_freq{0, r.location().y() + (int)nb_lines * char_height, 240, char_height};
+                painter.draw_rectangle(
+                    r_highlighted_freq,
+                    Color::white());
+            }
+            nb_lines++;
+        } else {
+            // rect is filled, we can break
+            break;
+        }
+    }
+    freqlist_nb_lines = nb_lines;
+    if (has_focus() || highlighted()) {
+        const Rect r_focus{r.left(), r.top(), r.width(), r.height()};
+        painter.draw_rectangle(
+            r_focus,
+            Color::white());
+    }
+}
+
+void FreqManUIList::set_db(freqman_db& db) {
+    freqlist_db = db;
+    current_index = 0;
+    highlighted_index = 0;
+}
+
+void FreqManUIList::on_hide() {
```

Comment:
nit: don't need to override these if you don't implement them.

Reply:
you're right, I've cleaned them out !